### ISSUE TITLE
inconsistent behavior: repeated reference to the same column versus row, dataset versus fast array

### ISSUE THREAD

**Issue by zhangyingmath** [rtosholdings/riptable#32]
import riptable as rt
rt.__version__
'1.0.19'

ds = rt.rt_dataset.Dataset({'f': [1,2,3,4], 'g': [5,6,7,8]})
ds
#	f	g
0	1	5
1	2	6
2	3	7
3	4	8

# repeated referencing column 'f' twice, but it only results once:
ds[:, ['f', 'g', 'f']]
#	f	g
0	1	5
1	2	6
2	3	7
3	4	8

fast_arr = ds.imatrix_make()

# in fast array, repeated reference will lead to this column getting duplicated. 
# this is also numpy and pandas behavior. 
fast_arr[:, [0,1,0]]
FastArray([[1, 5, 1],
           [2, 6, 2],
           [3, 7, 3],
           [4, 8, 4]])

# On the other hand, if I have repeated reference to the same row multiple times, then both dataset and fast array will duplicated it
ds[[0,1,0], :]

#	f	g
0	1	5
1	2	6
2	1	5

fast_arr[[0,1,0], :]
FastArray([[1, 5],
           [2, 6],
           [1, 5]])

**Comment by tdimitri** [comment]
A dataset is (in simple terms) a dictionary of FastArrays.  The dictionary only allows the column name to appear once.  This is different from a 2d matrix (which is what imatrix_make returns).  A 2d array (or matrix) does not have column names, and so it can duplicate columns.

So for ds[:, ['f', 'g', 'f']] to work, we would have to create another name for the second 'f' such as 'f_1'.  Is this acceptable?

**Comment by zhangyingmath** [comment]
I see, for now I don't have a strong opinion, especially if changing the behavior of this would have wide impact. But I would recommend we be more explicit about it than implicit. For example, if we decide to stick with this behavior, then we can raise an error as soon as the user asks for repeated columns.

**Comment by zhangyingmath** [comment]
Thanks for the discussion on Monday. As we discussed, the bracket operation can be one of a few potentially different getter functions for columns of a dataset. Suppose there is a dataset with only two columns (say 'f' and 'g'), but the user wants to access "three" columns (he asked for 'f' twice), I can think of four different behaviors:

1. Raise an exception. (I prefer this to be the default option, but I am happy to hear what others think.)
2. Give him two columns back, with no warning messages. (This is the current behavior.)
3. Give him three columns, 'f', 'f_2' and 'g', but 'f' and 'f_2' are simply views to the same underlying object in memory. (in other words, if you modify 'f_2', you will affect 'f'.)
4. Similar to 3, except 'f_2' is a copy of 'f', so modifying it does not affect 'f'. 

The follow code snippet (kind of) illustrates a behavior that resembles 4:

**Comment by zhangyingmath** [comment]
(have trouble editing my comment above, I will continue here, sorry for the truncated message):

Discussion:

The follow code snippet (kind of) illustrates a behavior that resembles 4: 

import numpy as np
a = np.array([[1,2], [4,5]])

a
array([[1, 2],
       [4, 5]])

b = a[:, [0, 1, 0]]

b
array([[1, 2, 1],
       [4, 5, 4]])

b[:, 2] = [1, 2]

b
array([[1, 2, 1],
       [4, 5, 2]])

Suppose I have a dataset, I can access this numpy API through imatrix_make:
ds_a = rt.Dataset({'f': [1, 4], 'g': [2, 5]})

ar_a = ds_a.imatrix_make()

ar_a
FastArray([[1, 2],
           [4, 5]])

ar_b = ar_a[:, [0, 1, 0]]

ar_b[:, 2] = [1, 2]

ar_b
FastArray([[1, 2, 1],
           [4, 5, 2]])

To summarize, I think my questions are: what should the default API be, and also should we offer more APIs (for example, should we offer one version that would eliminate the user's need to explicitly falling back to numpy by calling imatrix_make? After all, using named columns is easier than referring to numbered columns sometimes.)

**Comment by zhangyingmath** [comment]
Just realized imatrix_make doesn't like str columns, so my above approach wouldn't always work. Maybe it is worth its own API then.
